emscripten/emscripten_runner.js

```diff
@@ -0,0 +1,118 @@
+const { opendir } = require('node:fs/promises');
+const { loadPyodide } = require('pyodide');
+
+async function findWheel(distDir) {
+  const dir = await opendir(distDir);
+  for await (const dirent of dir) {
+    if (dirent.name.endsWith('wasm32.whl')) {
+      return dirent.name;
+    }
+  }
+}
+
+const pkgDir = process.argv[2];
+const distDir = pkgDir + '/dist';
+
+function make_tty_ops(stream){
+  return {
+    // get_char has 3 particular return values:
+    // a.) the next character represented as an integer
+    // b.) undefined to signal that no data is currently available
+    // c.) null to signal an EOF
+    get_char(tty) {
+      if (!tty.input.length) {
+        var result = null;
+        var BUFSIZE = 256;
+        var buf = Buffer.alloc(BUFSIZE);
+        var bytesRead = fs.readSync(process.stdin.fd, buf, 0, BUFSIZE, -1);
+        if (bytesRead === 0) {
+          return null;
+        }
+        result = buf.slice(0, bytesRead);
+        tty.input = Array.from(result);
+      }
+      return tty.input.shift();
+    },
+    put_char(tty, val) {
+      try {
+        if(val !== null){
+          tty.output.push(val);
+        }
+        if (val === null || val === 10) {
+          process.stdout.write(Buffer.from(tty.output));
+          tty.output = [];
+        }
+      } catch(e){
+        console.warn(e);
+      }
+    },
+    flush(tty) {
+      if (!tty.output || tty.output.length === 0) {
+        return;
+      }
+      stream.write(Buffer.from(tty.output));
+      tty.output = [];
+    }
+  };
+}
+
+function setupStreams(FS, TTY){
+  let mytty = FS.makedev(FS.createDevice.major++, 0);
```

Comment:
Cool, so did you manage to get live output from pytest with this?

Reply:
Yup, it almost works like a normal tty: it can update the same line, use color, etc. There is some trouble with flushing, I think the problem might be here:
https://github.com/python/cpython/blob/3.10/Modules/_io/textio.c#L1536-L1537
`sys.stdout.flush()` apparently doesn't actually flush file descriptor 1.

I think that a solution would be that shouldn't buffer writes to stdout.

It would be good to add something like this to Pyodide. Some issues:
1. By default, `sys.stdout.isatty()` is `True` but writing out ANSI color codes and terminal commands to `console.log` just prints nonsense. It should be false in these cases.
2. If you pass a custom `stdout` to `loadPyodide`, then `sys.stdout.isatty()` returns `False` and there is no way to ask it to return `True`.
3. Node applications can work correctly as ttys as (mostly) demonstrated here but it would be nice to offer that capability in some simple way for other node cli applications. Enabling this in tests breaks our node test runner though.

Most of these are also Emscripten issues. See related:
https://github.com/emscripten-core/emscripten/pull/17320